pylon-core/src/main/kotlin/io/github/pylonmc/pylon/core/entity/EntityStorage.kt

```diff
@@ -0,0 +1,162 @@
+    internal fun startAutosaveTask() {
+        Bukkit.getScheduler().runTaskTimer(pluginInstance, Runnable {
+            // TODO this saves all entities at once, potentially leading to large pauses
+            for (entity in entities.values) {
+                entity.write()
+            }
+        }, AUTOSAVE_INTERVAL_TICKS, AUTOSAVE_INTERVAL_TICKS)
+    }
+
+    @JvmStatic
+    fun get(uuid: UUID): PylonEntity<*, *>?
+        = lockEntityRead { entities[uuid] }
+
+    @JvmStatic
+    fun get(entity: Entity): PylonEntity<*, *>?
+        = get(entity.uniqueId)
+
+    @JvmStatic
+    fun <T> getAs(clazz: Class<T>, uuid: UUID): T? {
+        val entity = get(uuid) ?: return null
+        return clazz.cast(entity)
+    }
+
+    @JvmStatic
+    fun <T> getAs(clazz: Class<T>, entity: Entity): T?
+        = getAs(clazz, entity.uniqueId)
+
+    inline fun <reified T> getAs(uuid: UUID): T?
+        = getAs(T::class.java, uuid)
+
+    inline fun <reified T> getAs(entity: Entity): T?
+        = getAs(T::class.java, entity)
+
+    fun getByKey(key: NamespacedKey): Collection<PylonEntity<*, *>> =
+        if (PylonRegistry.ENTITIES.contains(key)) {
+            lockEntityRead {
+                entitiesByKey[key].orEmpty()
+            }
+        } else {
+            emptySet()
+        }
+
+    @JvmStatic
+    fun isPylonEntity(uuid: UUID): Boolean
+        = get(uuid) != null
+
+    @JvmStatic
+    fun isPylonEntity(entity: Entity): Boolean
+        = get(entity) != null
+
+    @JvmStatic
+    fun add(entity: PylonEntity<*, *>) = lockEntityWrite {
+        entities[entity.entity.uniqueId] = entity
+        entitiesByKey.getOrPut(entity.schema.key) { mutableSetOf() }.add(entity)
+    }
+
+    @EventHandler
+    private fun onEntityLoad(event: EntitiesLoadEvent) {
+        for (entity in event.entities) {
+            PylonEntity.deserialize(entity)?.let {
+                add(it)
+                PylonEntityLoadEvent(it).callEvent()
+            }
+        }
+    }
+
+    // This currently does not differentiate between unloaded and dead entities because the API
+    // is broken (lol), hence the lack of an entity death listener
+    @EventHandler
+    private fun onEntityUnload(event: EntityRemoveFromWorldEvent) {
+        val pylonEntity = get(event.getEntity().uniqueId) ?: return
+        PylonEntity.serialize(pylonEntity)
+        lockEntityWrite {
+            entities.remove(pylonEntity.entity.uniqueId)
+            entitiesByKey[pylonEntity.schema.key]!!.remove(pylonEntity)
+            if (entitiesByKey[pylonEntity.schema.key]!!.isEmpty()) {
+                entitiesByKey.remove(pylonEntity.schema.key)
+            }
+        }
+        PylonEntityUnloadEvent(pylonEntity).callEvent()
+    }
+
+    @EventHandler(ignoreCancelled = true)
+    private fun onEntityDeath(event: EntityDeathEvent)
+        = get(event.entity)?.let {
+            PylonEntityDeathEvent(it, event).callEvent()
+        }
```

Comment:
```suggestion
    private fun onEntityDeath(event: EntityDeathEvent) {
        get(event.entity)?.let {
            PylonEntityDeathEvent(it, event).callEvent()
        }
    }
```
paper doesn't like it when you return non-void from a evt listener, hence this warn when starting the server:
`[23:15:38 WARN]: @EventHandler method io.github.pylonmc.pylon.core.entity.EntityStorage#onEntityDeath returns non-void type java.lang.Boolean. This is unsupported behavior and will no longer work in a future version of Paper. This should be reported to the developers of PylonCore vSNAPSHOT ()`